### ISSUE TITLE
Tests are not building

### ISSUE THREAD

**Issue by 0nkery** [TyOverby/bincode#77]
Trying to run tests with nightly `rustc 1.11.0-nightly (bb4a79b08 2016-06-15)`.

``` shell
tests/test.rs:405:68: 405:68 error: mismatched types [E0308]
tests/test.rs:405     #[derive(RustcEncodable, RustcDecodable, Serialize, Deserialize, Debug)]
                                                                                     ^
tests/test.rs:405:57: 405:68 note: in this expansion of try! (defined in <std macros>)
tests/test.rs:405:57: 405:68 note: in this expansion of #[derive_Deserialize] (defined in tests/test.rs)
tests/test.rs:405:68: 405:68 help: run `rustc --explain E0308` to see a detailed explanation
tests/test.rs:405:68: 405:68 note: expected type `serde::Deserialize`
tests/test.rs:405:68: 405:68 note:    found type `serde::Deserialize`
tests/test.rs:405:57: 405:68 note: the lifetime 'a as defined on the block at 405:56...
tests/test.rs:405     #[derive(RustcEncodable, RustcDecodable, Serialize, Deserialize, Debug)]
                                                                          ^~~~~~~~~~~
tests/test.rs:405:68: 405:68 note: ...does not necessarily outlive the static lifetime
```

As far as I can tell this error points `RefBox`'s lifetimes.

**Comment by TyOverby** [comment]
@0nkery @dtolnay @erickt: could this be a change in the Deserialize codegen?  Could one of the serde maintainers tell us if there were any changes to codegen that would break us like this?

**Comment by TyOverby** [comment]
Until this is figured out, I'll be disabling the tests over in https://github.com/TyOverby/bincode/pull/78

**Comment by dtolnay** [comment]
So far I have narrowed it down to the following code does not compile:

```rust
#[derive(Deserialize)]
struct Message<'a>(RefBox<'a, ()>);
```

Now I will start looking at the generated code to find the problem.

**Comment by dtolnay** [comment]
This failure was correct on Serde's part and a bug in bincode. The code was trying to derive Deserialize for a type containing `RefBox<'a, _>`, but that does not implement Deserialize because only `RefBox<'static, _>` implements Deserialize. I have a fix in #79.
